File: Olympiad/COCI/official/2009/contest2_solutions/perket.pas
program perket;

const MAXN = 10;

var
   i, n : longint;
   k, g : array[1..MAXN] of longint;
   best : longint;

function diff( x, y : longint ) : longint;
begin
   if x < y then diff := y-x else diff := x-y;
end;

procedure rek( i, kis, gor : longint );
begin
   if i > n then begin
      if (gor > 0) and (diff(kis,gor) < best) then best := diff(kis,gor);
   end else begin
      rek( i+1, kis, gor );
      rek( i+1, kis*k[i], gor+g[i] );
   end;
end;

begin
   readln( n );
   for i := 1 to n do readln( k[i], g[i] );
   
   best := 1000000000;
   rek( 1, 1, 0 );

   writeln( best );   
end.

File: Olympiad/COCI/official/2009/contest2_solutions/perket.cpp
#include <cstdio>

#define MAXN 10

int n, k[MAXN], g[MAXN], best = 1000000000;

int diff( int x, int y ) { return x<y ? y-x : x-y; }

void rek( int i, int kis, int gor ) {
   if( i == n ) {
      if( gor > 0 && diff(kis,gor) < best ) best = diff(kis,gor);
   } else {
      rek( i+1, kis, gor );
      rek( i+1, kis*k[i], gor+g[i] );
   }
}

int main(void) {
   scanf( "%d", &n );
   for( int i = 0; i < n; ++i ) scanf( "%d%d", &k[i], &g[i] );

   rek( 0, 1, 0 );

   printf( "%d\n", best );
   return 0;
}
